### PR TITLE
V0.17.0 bump

### PR DESCRIPTION
**Pull request by antalszava**

### CODE REVIEW

**Comment by antalszava** [comment]
[ch7892]

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -1,32 +1,6 @@
...
-  Building a PennyLane image with the Qiskit plugin (change `plugin-name` for other plugins):
-  ```
-  make -f docker/Makefile build-plugin plugin-name=qiskit
-  ```
-
-  Building the PennyLane-QChem image:
-  ```
-  make -f docker/Makefile build-qchem
-  ```
+<h4>Circuit optimization</h4>
```
Comment (josh146) [inline .github/CHANGELOG.md:3]:
```suggestion
<h3>New features since the last release</h3>

<h4>Circuit optimization</h4>
```

File: .github/CHANGELOG.md [hunk 2]
```diff
@@ -83,6 +57,245 @@
    2: ──H──╰C────────RZ(0.4)──RZ(-0.4)──RX(0.3)──╰CY──Y──┤
   ```
 
+* A new quantum function transform has been added to push commuting
+  single-qubit gates through controlled operations.
+  [(#1464)](https://github.com/PennyLaneAI/pennylane/pull/1464)
```
Comment (josh146) [inline .github/CHANGELOG.md:62]: @antalszava @co9olguy I am wondering if having examples of every compilation transform might be information overload? An alternative is to keep the example for `qml.compile`, but replace the other examples with a condensed bullet list:

```markdown
  The following compilation transforms have been added and are also available
  to use, either independently, or within a `qml.compile` pipeline:

  * `commute_controlled`: push commuting single-qubit gates through controlled operations.
    [(#1464)](https://github.com/PennyLaneAI/pennylane/pull/1464)
    
  * `cancel_inverses`: removes adjacent pairs of operations that cancel out.
    [(#1455)](https://github.com/PennyLaneAI/pennylane/pull/1455)

  * `merge_rotations`: combines adjacent rotation gates of
    the same type into a single gate, including controlled rotations.
    [(#1455)](https://github.com/PennyLaneAI/pennylane/pull/1455)

  * `single_qubit_fusion`: acts on all sequences of
    single-qubit operations in a quantum function, and converts each
    sequence to a single `Rot` gate.
    [(#1458)](https://github.com/PennyLaneAI/pennylane/pull/1458)

  For more details on `qml.compile` and the available compilation transforms, see
  [the compilation documentation](https://pennylane.readthedocs.io/en/latest/code/qml_transforms.html#transforms-for-circuit-compilation).
```

File: .github/CHANGELOG.md [hunk 3]
```diff
@@ -83,6 +57,245 @@
...
+
+  ```pycon
+  >>> optimized_circuit = qml.transforms.commute_controlled(direction="right")(circuit)
+  >>> dev = qml.device('default.qubit', wires=3)
+  >>> qnode = qml.QNode(optimized_circuit, dev)
+  >>> print(qml.draw(qnode)(0.5))
+   0: ──╭C──╭C──S──Rϕ(0.25)──T──┤ ⟨Z⟩
+   1: ──╰X──├C──────────────────┤
+   2: ──────╰X──X───────────────┤
+  ```
```
Comment (josh146) [inline .github/CHANGELOG.md:85]: For the compilation transforms, how do we feel about using decorator syntax instead?

```python
dev = qml.device('default.qubit', wires=3)

@qml.qnode(dev)
@qml.transforms.commute_controlled(direction="right")
def circuit(theta):
    qml.PauliX(wires=2)
    qml.S(wires=0)
    qml.CNOT(wires=[0, 1])
    qml.PhaseShift(theta/2, wires=0)
    qml.T(wires=0)
    qml.Toffoli(wires=[0, 1, 2])
    return qml.expval(qml.PauliZ(0))

>>> print(qml.draw(circuit)(0.5))
 0: ──╭C──╭C──S──Rϕ(0.25)──T──┤ ⟨Z⟩
 1: ──╰X──├C──────────────────┤
 2: ──────╰X──X───────────────┤
```

To me, it makes the examples easier to read, and slightly shorter (there is less duplication).

File: .github/CHANGELOG.md [hunk 4]
```diff
@@ -1,32 +1,6 @@
-# Release 0.17.0-dev (development release)
+# Release 0.17.0 (current release)
```
Comment (josh146) [inline .github/CHANGELOG.md:1]: @antalszava I can't comment on it, but beware of accidental double-backticks :) I can see one down on line 44

image

File: .github/CHANGELOG.md [hunk 5]
```diff
@@ -83,6 +57,245 @@
...
+   0: ──Rot(3.24, 1.57, 0)──╭Z──Rot(2.36, 1.57, -1.57)────┤ ⟨X⟩
+   1: ──Rot(3.14, 0.2, 0)───╰C──Rot(0.406, 0.382, 0.406)──┤
+  ```
+
+<h4>Improved Hamiltonian simulations</h4>
+
+* Added a sparse Hamiltonian observable and the functionality to support computing its expectation
+  value. [(#1398)](https://github.com/PennyLaneAI/pennylane/pull/1398)
+
+  For example, the following QNode returns the expectation value of a sparse Hamiltonian:
```
Comment (josh146) [inline .github/CHANGELOG.md:180]: I might be careful with this one:

- Only `default.qubit` is supported at the moment. However the entry makes it sound like it leads to faster execution in general.
- Only expvals are support, I believe(?)

Finally, I think the plan is to replace it with `qml.expval(qml.Hamiltonian(...))`, and simply use the fact that `qml.Hamiltonian` is _already_ a sparse representation of the Hamiltonian! So not sure how much we want to publicize this, since this change will likely be in the next release.

@mariaschuld, any thoughts here?

Comment (josh146) [reply .github/CHANGELOG.md:180]: Perhaps it should be sufficient just to move this category 'lower' down the list of new features

File: .github/CHANGELOG.md [hunk 6]
```diff
@@ -83,6 +57,245 @@
...
+
+  ``` python
+  dev = qml.device("default.qubit", wires=2)
+  H = qml.Hamiltonian([1., 2., 3.],  [qml.PauliZ(0), qml.PauliY(0), qml.PauliZ(1)])
+  w = qml.init.strong_ent_layers_uniform(1, 2, seed=1967)
+
+  @qml.qnode(dev)
+  def circuit(w):
+      qml.templates.StronglyEntanglingLayers(w, wires=range(2))
+      return qml.expval(H)
```
Comment (josh146) [inline .github/CHANGELOG.md:214]: @mariaschuld just double checking here as well - I remember we had a conversation regarding whether this should be advertised now or in v0.18

File: .github/CHANGELOG.md [hunk 7]
```diff
@@ -83,6 +57,245 @@
...
+  [ 1.02348685e-16 -7.77143238e-01 -1.74708049e-01]]]
+  ```
+
+  Note that other measurement types like `var(H)` or `sample(H)`, as well
+  as multiple expectations like `expval(H1), expval(H2)` are not supported.
+
+* Added functionality to compute the sparse matrix representation of a `qml.Hamiltonian` object.
+  [(#1394)](https://github.com/PennyLaneAI/pennylane/pull/1394)
+
+<h4> QNodes are more convenient</h4>
```
Comment (josh146) [inline .github/CHANGELOG.md:231]: Trying to think of an alternative here that might be more descriptive 🤔

File: .github/CHANGELOG.md [hunk 8]
```diff
@@ -83,6 +57,245 @@
...
+
+* Added functionality to `qml.sample()` to extract samples from the basis states of
+  the device (currently only for qubit devices). Additionally, `wires` can be
+  specified to only return samples from those wires.
+  [(#1441)](https://github.com/PennyLaneAI/pennylane/pull/1441)
```
Comment (josh146) [inline .github/CHANGELOG.md:236]:
```suggestion
* Computational basis samples directly from the underlying device can
  now be returned directly from QNodes via `qml.sample()`.
  [(#1441)](https://github.com/PennyLaneAI/pennylane/pull/1441)
```

Comment (josh146) [reply .github/CHANGELOG.md:236]: This reminds me: there are several demos we should update to use this :) E.g., the [QAOA demo](https://pennylane.ai/qml/demos/tutorial_qaoa_maxcut.html) comes to mind (I think there are others).

Currently, existing demos use two separate tricks to get the computational basis samples:

1. They use `qml.Hermitian`:

   ```python
   return qml.sample(qml.Hermitian(np.diag(2 ** np.arange(num_wires), wires=wires))

   res = qnode(params) # returns list of state samples in base-10
   output_samples = convert_to_binary(res) # need to convert to binary representation
   ```

2. Or, they use

   ```python
    return [qml.sample(qml.PauliZ(w)) for w in wires]

    res = qnode(params) # returns state samples using -1, 1
    output_samples = (res + 1) / 2 # convert to using 0, 1
    ```

Both of these approaches add overhead/complexity (_especially_ number 1!) and so should be updated. It also helps propagate to our users this new functionality.

File: .github/CHANGELOG.md [hunk 9]
```diff
@@ -83,6 +57,245 @@
...
+  ```
+
+  ```pycon
+  >>> print(qml.draw(circuit)(0.6))
+  0: ──RY(0.6)──RX(0.4)──┤ ⟨Z⟩
+  ```
+
+  Previously instantiated measurements can also be applied to QNodes.
+
+<h4>Resource Devicer Tracker</h4>
```
Comment (josh146) [inline .github/CHANGELOG.md:297]: There is a rendering issue with the device tracker below:

image

File: .github/CHANGELOG.md [hunk 10]
```diff
@@ -128,30 +341,35 @@
...
-  def circuit(w):
-      qml.templates.StronglyEntanglingLayers(w, wires=range(2))
-      return qml.expval(H)
+* Docker support for building PennyLane with support for all interfaces (TensorFlow,
+  Torch, and Jax), as well as device plugins and QChem, for GPUs and CPUs, has been added.
+  [(#1372)](https://github.com/PennyLaneAI/pennylane/issues/1372)
+
+  The build process using Docker and Makefile works as follows:
+
+  Building a core PennyLane image:
```
Comment (josh146) [inline .github/CHANGELOG.md:352]: Seems like these should be bullet points?

File: .github/CHANGELOG.md [hunk 11]
```diff
@@ -182,32 +400,7 @@
...
-  ```pycon
-  >>> optimized_circuit = qml.transforms.commute_controlled(direction="right")(circuit)
-  >>> dev = qml.device('default.qubit', wires=3)
-  >>> qnode = qml.QNode(optimized_circuit, dev)
-  >>> print(qml.draw(qnode)(0.5))
-   0: ──╭C──╭C──S──Rϕ(0.25)──T──┤ ⟨Z⟩
-   1: ──╰X──├C──────────────────┤
-   2: ──────╰X──X───────────────┤
-  ```
+<h4>Extended operations and templates</h4>
```
Comment (josh146) [inline .github/CHANGELOG.md:403]:
```suggestion
<h4>Even more new operations and templates</h4>
```

Comment (josh146) [reply .github/CHANGELOG.md:403]: > A decomposition has been added to QubitUnitary that makes the single-qubit case fully differentiable in all interfaces. 

This one is really cool! Should we move it higher?

**Review by josh146 — CHANGES_REQUESTED** [review]
Thanks @antalszava! I've done a first pass, mostly looking at high-level stuff such as rendering, flow, structure, and order 🙂

File: .github/CHANGELOG.md [hunk 12]
```diff
@@ -492,7 +533,7 @@
 * Change the order of the covariance matrix and the vector of means internally
   in `default.gaussian`. [(#1331)](https://github.com/PennyLaneAI/pennylane/pull/1331)
 
-* Added the `id` attribute to templates, which was missing from 
+* Added the `id` attribute to templates, which was missing from
   PR [(#1377)](https://github.com/PennyLaneAI/pennylane/pull/1377).
   [(#1438)](https://github.com/PennyLaneAI/pennylane/pull/1438)
```
Comment (co9olguy) [inline .github/CHANGELOG.md:538]:
```suggestion
* Added the `id` attribute to templates). [(#1438)](https://github.com/PennyLaneAI/pennylane/pull/1438)
```

**Review by co9olguy — APPROVED** [review]
Approved subject to @josh146's suggested changes. I think this is in pretty great shape :+1:

File: .github/CHANGELOG.md (hunk 4 above)
Comment (antalszava) [reply .github/CHANGELOG.md:1]: Updated! How come it's good to avoid them? :thinking: Would have thought both single and double-backticks work.

File: .github/CHANGELOG.md (hunk 5 above)
Comment (antalszava) [reply .github/CHANGELOG.md]: Sounds good! Moved it lower and added a note on `default.qubit`.

File: .github/CHANGELOG.md (hunk 8 above)
Comment (antalszava) [reply .github/CHANGELOG.md:88]: Made a note!

File: .github/CHANGELOG.md (hunk 2 above)
Comment (co9olguy) [reply .github/CHANGELOG.md]: yup I agree we can condense the story around compilation

File: .github/CHANGELOG.md (hunk 3 above)
Comment (co9olguy) [reply .github/CHANGELOG.md]: :+1:

File: .github/CHANGELOG.md (hunk 10 above)
Comment (antalszava) [reply .github/CHANGELOG.md]: Good catch! Updated

File: .github/CHANGELOG.md (hunk 9 above)
Comment (antalszava) [reply .github/CHANGELOG.md:148]: Good catch! Changed it.

File: .github/CHANGELOG.md [hunk 13]
```diff
@@ -492,7 +533,7 @@
...
```
[the remark this answers is not in the document]
Comment (antalszava) [reply .github/CHANGELOG.md:444]: Thanks! Updated.

File: .github/CHANGELOG.md (hunk 3 above)
Comment (antalszava) [reply .github/CHANGELOG.md]: With the condensed bullet list of compilation transforms this example has been removed.

File: .github/CHANGELOG.md (hunk 7 above)
Comment (antalszava) [reply .github/CHANGELOG.md:83]: Maybe "More convenient QNode usage"? I think both items here serve the purpose of user convenience. Therefore that might be a point worth having in the section name.

File: .github/CHANGELOG.md (hunk 11 above)
Comment (antalszava) [reply .github/CHANGELOG.md]: Would you suggest moving it higher within the "Even more new operations and templates" section of outside of it even?

File: .github/CHANGELOG.md (hunk 3 above)
Comment (josh146) [reply .github/CHANGELOG.md]: @antalszava there is the compilation example on lines 15-43 that can be updated to use decorator syntax :)

File: .github/CHANGELOG.md (hunk 4 above)
Comment (josh146) [reply .github/CHANGELOG.md:1]: Oh - Markdown only supports single backticks for code :) Double backticks rendering correctly is accident of some renderers, and is not always guaranteed to work.

File: .github/CHANGELOG.md (hunk 7 above)
Comment (josh146) [reply .github/CHANGELOG.md]: "QNodes are even more powerful"?

File: .github/CHANGELOG.md (hunk 11 above)
Comment (josh146) [reply .github/CHANGELOG.md]: I was thinking more moving it higher inside the "new operations" section :)

This one is something a lot of users have asked about (training qubit unitary), so is a nice one to make more visible

File: .github/CHANGELOG.md [hunk 14]
```diff
@@ -1,42 +1,22 @@
...
   [(#1475)](https://github.com/PennyLaneAI/pennylane/pull/1475)
 
-  For example, take the following quantum function:
+  For example, take the following decorated quantum function:
 
   ```python
+  dev = qml.device('default.qubit', wires=[0, 1, 2])
+
+  @qml.qnode(dev)
+  @qml.transforms.compile()
```
Comment (antalszava) [inline .github/CHANGELOG.md:19]:
```suggestion
  @qml.compile()
```

File: .github/CHANGELOG.md (hunk 9 above)
Comment (co9olguy) [reply .github/CHANGELOG.md:149]:
```suggestion
<h4>Resource Device Tracker</h4>
```

File: .github/CHANGELOG.md [hunk 15]
```diff
@@ -126,32 +191,94 @@
...
+  Total shots:  300
+  ```
+
+<h4>Containerization support</h4>
+
+* Docker support for building PennyLane with support for all interfaces (TensorFlow,
+  Torch, and Jax), as well as device plugins and QChem, for GPUs and CPUs, has been added.
+  [(#1391)](https://github.com/PennyLaneAI/pennylane/pull/1391)
+
+  The build process using Docker and Makefile works as follows:
```
Comment (josh146) [inline .github/CHANGELOG.md:203]:
```suggestion
  The build process using Docker and `make` requires that the repository source code
  is cloned or downloaded from GitHub, and provides the following options:
```

Comment (josh146) [reply .github/CHANGELOG.md:203]: @antalszava I just noticed something; the following instructions are also provided in the README, however they shouldn't be there - we made a concerted effort to simplify the README a couple of releases ago!

I recommend the following:

- Remove the docker section from the README
- Instead, add it to the developer guide as a new card here: https://pennylane.readthedocs.io/en/latest/development/guide.html
- Add a short link in the readme
- Add a short link here in the changelog, and remove the full list of commands

**Review by josh146 — APPROVED** [review]
Nice work @antalszava, a very nicely written and exciting changelog 🤩 

I left one suggestion re: the docker commands, but happy to approve (I would rather they be removed from the README)

File: .github/CHANGELOG.md (hunk 15 above)
Comment (antalszava) [reply .github/CHANGELOG.md]: Good point! Removed the docker section from the README. It also seems like we could add the description about Docker to the installation parts of the development guide and the user installation section. Let me open a separate PR.
